r2/service.go

```diff
@@ -59,86 +64,480 @@ func NewService(iOpts instrument.Options, rootPrefix string, store Store) server
 	return &service{iOpts: iOpts, store: store, rootPrefix: rootPrefix}
 }
 
+type payload struct {
+	data   []byte
+	status int
+}
+
+func newPayload(d []byte, status int) payload {
+	return payload{data: d, status: status}
+}
+
+func (p payload) write(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(p.status)
+	w.Write(p.data)
+}
+
+func writeAPIResponse(w http.ResponseWriter, code int, msg string) {
+	j, _ := json.Marshal(apiResponse{Code: code, Message: msg})
```

Comment:
nit: don't drop err here. can do a `panic` if you're sure this should never fail